packages/contract/src/builder.test-d.ts

```diff
@@ -0,0 +1,289 @@
+import {
+  defineConsumer,
+  defineContract,
+  defineExchange,
+  defineMessage,
+  definePublisher,
+  defineQueue,
+  defineQueueBinding,
+  mergeContracts,
+} from "./builder.js";
+import { describe, expectTypeOf, it } from "vitest";
+import type { MergeContracts } from "./types.js";
+import { z } from "zod";
+
+describe("mergeContracts - Type Tests", () => {
+  describe("Type safety and inference", () => {
+    it("should merge two contracts with publishers", () => {
+      // GIVEN
+      const orderExchange = defineExchange("orders", "topic", { durable: true });
+      const paymentExchange = defineExchange("payments", "topic", { durable: true });
+
+      const orderMessage = defineMessage(z.object({ orderId: z.string() }));
+      const paymentMessage = defineMessage(z.object({ paymentId: z.string() }));
+
+      const orderContract = defineContract({
+        exchanges: { orders: orderExchange },
+        publishers: {
+          orderCreated: definePublisher(orderExchange, orderMessage, {
+            routingKey: "order.created",
+          }),
+        },
+      });
+
+      const paymentContract = defineContract({
+        exchanges: { payments: paymentExchange },
+        publishers: {
+          paymentReceived: definePublisher(paymentExchange, paymentMessage, {
+            routingKey: "payment.received",
+          }),
+        },
+      });
+
+      // WHEN
+      const merged = mergeContracts(orderContract, paymentContract);
+
+      // THEN - Verify merged type matches expected structure
+      type Expected = MergeContracts<[typeof orderContract, typeof paymentContract]>;
+      expectTypeOf(merged).toMatchTypeOf<Expected>();
```

Comment:
no, your are expecting that what you expect is what you expect

you must write precisely what you expect

do it everywhere